File: pdfbridge/ionpair.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright (C) 2015 The ProteinDF development team.
# see also AUTHORS and README if provided.
#
# This file is a part of the ProteinDF software package.
#
# The ProteinDF is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# The ProteinDF is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with ProteinDF.  If not, see <http://www.gnu.org/licenses/>.

import pdfbridge
import logging
logger = logging.getLogger(__name__)

class IonPair(object):
    def __init__(self, model):
        self._model = pdfbridge.AtomGroup(model)

    def get_ion_pairs(self):
        ion_pairs = []
        (anion_list, cation_list) = self._get_ion_list()
        for anion_path, anion_pos_array in anion_list.items():
            for (anion_pos, anion_type) in anion_pos_array:
                
                for cation_path, cation_pos_array in cation_list.items():
                    for (cation_pos,cation_type) in cation_pos_array:

                        d = anion_pos.distance_from(cation_pos)
                        if d < 4.0:
                            logger.info("found ion pair: {}".format(anion_path))
                            logger.info("              : {}".format(cation_path))
                            ion_pairs.append((anion_path, cation_path, anion_type, cation_type))

        return ion_pairs
                    
    def _get_ion_list(self):
        anion_list = {}
        cation_list = {}

        for chain_key, chain in self._model.groups():
            for res_key, res in chain.groups():
                name = res.name

                if name == 'GLU':
                    anion_list[res.path] = [(self._get_center_GLU(res), 'GLU')]
                elif name == 'ASP':
                    anion_list[res.path] = [(self._get_center_ASP(res), 'ASP')]
                elif name == 'LYS':
                    cation_list[res.path] = [(self._get_center_LYS(res), 'LYS')]
                elif name == 'ARG':
                    cation_list[res.path] = [(self._get_center_ARG(res, 0), 'ARG'),
                                             (self._get_center_ARG(res, 1), 'ARG1'),
                                             (self._get_center_ARG(res, 2), 'ARG2')]

                if pdfbridge.AminoAcid.is_aminoacid(res):
                    if res.has_atom('H3'):
                        cation_list.setdefault(res.path, [])
                        cation_list[res.path].append((self._get_center_Nterm(res), 'NTM'))
                    if res.has_atom('OXT'):
                        anion_list.setdefault(res.path, [])
                        anion_list[res.path].append((self._get_center_Cterm(res), 'CTM'))

        return (anion_list, cation_list)
                        
        
    def _get_center_Nterm(self, res):
        """
        N末端のイオン対判定用座標を返す
        """
        return res['N'].xyz

        
    def _get_center_Cterm(self, res):
        """
        C末端のイオン対判定用座標を返す
        """
        ag = pdfbridge.AtomGroup()
        ag.set_atom('C', res['C'])
        ag.set_atom('O1', res['O'])
        ag.set_atom('O2', res['OXT'])
        return ag.center()
        

    def _get_center_GLU(self, res):
        """
        GLUのイオン対判定用座標を返す
        """
        ag = pdfbridge.AtomGroup()
        ag.set_atom('C', res['CD'])
        ag.set_atom('O1', res['OE1'])
        ag.set_atom('O2', res['OE2'])
        return ag.center()

        
    def _get_center_ASP(self, res):
        """
        ASPのイオン対判定用座標を返す
        """
        ag = pdfbridge.AtomGroup()
        ag.set_atom('C', res['CG'])
        ag.set_atom('O1', res['OD1'])
        ag.set_atom('O2', res['OD2'])
        return ag.center()
        

    def _get_center_LYS(self, res):
        """
        LYSのイオン対判定用座標を返す
        """
        return res['NZ'].xyz


    def _get_center_ARG(self, res, case=0):
        """
        case: 0; 中央
        case: 1; NH1側
        case: 2; NH2側
        """
        case = int(case)

        answer = pdfbridge.Position()
        ag = pdfbridge.AtomGroup()
        if case == 0:
            ag.set_atom('NH1', res['NH1'])
            ag.set_atom('NH2', res['NH2'])
            ag.set_atom('CZ', res['CZ'])
            answer = ag.center()
        elif case == 1:
            answer = res['NH1'].xyz
        elif case == 2:
            answer = res['NH2'].xyz
        else:
            logger.warning("unknown mode={}".format(mode))
            
        return answer


if __name__ == "__main__":
    main()

File: rust/crates/proteindf-bridge/src/ion_pair.rs
// SPDX-FileCopyrightText: The ProteinDF development team
// SPDX-License-Identifier: GPL-3.0-or-later

use crate::amino_acid::AminoAcid;
use crate::atom_group::AtomGroup;
use crate::position::Position;

type ResidueSites = Vec<(String, Vec<(Position, String)>)>;

/// Represents a detected ion-pair interaction between an anion and a cation site.
#[derive(Debug, Clone, PartialEq)]
pub struct IonPairRecord {
    pub anion_path: String,
    pub cation_path: String,
    pub anion_type: String,
    pub cation_type: String,
}

impl IonPairRecord {
    pub fn new(
        anion_path: impl Into<String>,
        cation_path: impl Into<String>,
        anion_type: impl Into<String>,
        cation_type: impl Into<String>,
    ) -> Self {
        Self {
            anion_path: anion_path.into(),
            cation_path: cation_path.into(),
            anion_type: anion_type.into(),
            cation_type: cation_type.into(),
        }
    }

    /// Converts this record into a tuple `(anion_path, cation_path, anion_type, cation_type)` matching Python.
    pub fn into_tuple(self) -> (String, String, String, String) {
        (
            self.anion_path,
            self.cation_path,
            self.anion_type,
            self.cation_type,
        )
    }
}

/// Detects salt bridges and ion-pair interactions in protein models corresponding to `proteindf_bridge.ionpair.IonPair`.
#[derive(Debug, Clone)]
pub struct IonPair {
    model: AtomGroup,
}

impl IonPair {
    /// Distance threshold for ion-pair interaction (4.0 Å).
    pub const MAX_DISTANCE: f64 = 4.0;

    /// Creates a new `IonPair` detector for the given model.
    pub fn new(model: &AtomGroup) -> Self {
        Self {
            model: model.clone(),
        }
    }

    /// Finds and returns all ion-pair interactions within `MAX_DISTANCE` (4.0 Å).
    pub fn get_ion_pairs(&self) -> Vec<IonPairRecord> {
        let mut ion_pairs = Vec::new();
        let (anion_list, cation_list) = self.get_ion_list();

        for (anion_path, anion_pos_array) in &anion_list {
            for (anion_pos, anion_type) in anion_pos_array {
                for (cation_path, cation_pos_array) in &cation_list {
                    for (cation_pos, cation_type) in cation_pos_array {
                        let d = anion_pos.distance_from(cation_pos);
                        if d < Self::MAX_DISTANCE {
                            ion_pairs.push(IonPairRecord::new(
                                anion_path,
                                cation_path,
                                anion_type,
                                cation_type,
                            ));
                        }
                    }
                }
            }
        }

        ion_pairs
    }

    /// Convenience static helper to detect ion pairs directly from an `AtomGroup`.
    pub fn find_ion_pairs(model: &AtomGroup) -> Vec<IonPairRecord> {
        let detector = Self::new(model);
        detector.get_ion_pairs()
    }

    /// Extracts anion and cation interaction sites from the structure.
    ///
    /// NOTE on divergence from Python:
    /// In the original Python implementation (`ionpair.py`), missing expected atoms
    /// (e.g., missing `OE2` in GLU or `CZ` in ARG) raises a `KeyError` during dictionary
    /// lookup, which crashes the entire ion-pair analysis.
    /// In this Rust implementation, `get_center_*` functions return `Option<Position>`.
    /// When any required atom is missing, the incomplete residue site is safely skipped
    /// (`None`), allowing the analysis of other intact residues in the model to proceed
    /// without crashing.
    fn get_ion_list(&self) -> (ResidueSites, ResidueSites) {
        let mut anion_list: ResidueSites = Vec::new();
        let mut cation_list: ResidueSites = Vec::new();

        for (_chain_key, chain) in self.model.groups() {
            for (_res_key, res) in chain.groups() {
                let name = res.name.trim();

                if name == "GLU" {
                    if let Some(pos) = Self::get_center_glu(res) {
                        anion_list.push((res.path().to_string(), vec![(pos, "GLU".to_string())]));
                    }
                } else if name == "ASP" {
                    if let Some(pos) = Self::get_center_asp(res) {
                        anion_list.push((res.path().to_string(), vec![(pos, "ASP".to_string())]));
                    }
                } else if name == "LYS" {
                    if let Some(pos) = Self::get_center_lys(res) {
                        cation_list.push((res.path().to_string(), vec![(pos, "LYS".to_string())]));
                    }
                } else if name == "ARG" {
                    let mut arg_sites = Vec::new();
                    if let Some(pos0) = Self::get_center_arg(res, 0) {
                        arg_sites.push((pos0, "ARG".to_string()));
                    }
                    if let Some(pos1) = Self::get_center_arg(res, 1) {
                        arg_sites.push((pos1, "ARG1".to_string()));
                    }
                    if let Some(pos2) = Self::get_center_arg(res, 2) {
                        arg_sites.push((pos2, "ARG2".to_string()));
                    }
                    if !arg_sites.is_empty() {
                        cation_list.push((res.path().to_string(), arg_sites));
                    }
                }

                if AminoAcid::is_aminoacid(res) {
                    if res.has_atom("H3") {
                        if let Some(pos) = Self::get_center_nterm(res) {
                            if let Some(entry) =
                                cation_list.iter_mut().find(|(p, _)| p == res.path())
                            {
                                entry.1.push((pos, "NTM".to_string()));
                            } else {
                                cation_list
                                    .push((res.path().to_string(), vec![(pos, "NTM".to_string())]));
                            }
                        }
                    }
                    if res.has_atom("OXT") {
                        if let Some(pos) = Self::get_center_cterm(res) {
                            if let Some(entry) =
                                anion_list.iter_mut().find(|(p, _)| p == res.path())
                            {
                                entry.1.push((pos, "CTM".to_string()));
                            } else {
                                anion_list
                                    .push((res.path().to_string(), vec![(pos, "CTM".to_string())]));
                            }
                        }
                    }
                }
            }
        }

        (anion_list, cation_list)
    }

    /// Returns the position of N if present; returns None if missing.
    fn get_center_nterm(res: &AtomGroup) -> Option<Position> {
        res.get_atom("N").map(|a| a.xyz)
    }

    /// Computes the geometric center of C, O, OXT. Returns None if any atom is missing.
    fn get_center_cterm(res: &AtomGroup) -> Option<Position> {
        let c = res.get_atom("C")?;
        let o = res.get_atom("O")?;
        let oxt = res.get_atom("OXT")?;
        Some((c.xyz + o.xyz + oxt.xyz) / 3.0)
    }

    /// Computes the geometric center of CD, OE1, OE2. Returns None if any atom is missing.
    fn get_center_glu(res: &AtomGroup) -> Option<Position> {
        let cd = res.get_atom("CD")?;
        let oe1 = res.get_atom("OE1")?;
        let oe2 = res.get_atom("OE2")?;
        Some((cd.xyz + oe1.xyz + oe2.xyz) / 3.0)
    }

    /// Computes the geometric center of CG, OD1, OD2. Returns None if any atom is missing.
    fn get_center_asp(res: &AtomGroup) -> Option<Position> {
        let cg = res.get_atom("CG")?;
        let od1 = res.get_atom("OD1")?;
        let od2 = res.get_atom("OD2")?;
        Some((cg.xyz + od1.xyz + od2.xyz) / 3.0)
    }

    /// Returns the position of NZ if present; returns None if missing.
    fn get_center_lys(res: &AtomGroup) -> Option<Position> {
        res.get_atom("NZ").map(|a| a.xyz)
    }

    /// Computes ARG site position (case 0: center of NH1, NH2, CZ; case 1: NH1; case 2: NH2).
    /// Returns None if any required atom is missing.
    fn get_center_arg(res: &AtomGroup, case: usize) -> Option<Position> {
        match case {
            0 => {
                let nh1 = res.get_atom("NH1")?;
                let nh2 = res.get_atom("NH2")?;
                let cz = res.get_atom("CZ")?;
                Some((nh1.xyz + nh2.xyz + cz.xyz) / 3.0)
            }
            1 => res.get_atom("NH1").map(|a| a.xyz),
            2 => res.get_atom("NH2").map(|a| a.xyz),
            _ => None,
        }
    }
}
